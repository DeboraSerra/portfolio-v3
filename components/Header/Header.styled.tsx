import styled from "styled-components";

export const Header = styled.header`
  position: relative;
  background-color: ${(props) => props.theme.bg.secondary};
  padding: 12px 24px;
  box-shadow: 0 4px 24px ${(props) => props.theme.bg.input};
  height: 64px;

  .container {
    display: flex;
    justify-content: space-between;
    align-items: center;

    .header {
      &__title {
        font-size: ${(props) => props.theme.fontSizes[5]};
        font-weight: ${(props) => props.theme.fontWeights.heading};
        color: ${(props) => props.theme.text.primary};
        text-decoration: none;
      } // &__title

      &__links {
        display: flex;
        flex-grow: 1;
        justify-content: space-around;
        align-items: center;
        max-width: 700px;
      } // &__links

      &__link {
        position: relative;
        display: flex;
        justify-content: center;
        align-items: center;

        color: ${(props) => props.theme.text.secondary};
        cursor: pointer;

        &:hover {
          color: ${(props) => props.theme.text.tertiary};
          text-decoration: underline;
        } // &:hover
      } // &__link

      &__theme-btn,
      &__nav-btn {
        color: ${(props) => props.theme.text.secondary};
        cursor: pointer;
        font-size: ${(props) => props.theme.fontSizes[3]};
      } // &__theme-btn

      @media only screen and (max-width: 576px) {
        &__links {
          position: absolute;
          top: 64px;
          left: 0;
          display: flex;
          justify-content: space-around;
          width: 100%;
          background-color: ${(props) => props.theme.bg.secondary};
          padding: 12px 24px;
          box-shadow: 0 12px 24px ${(props) => props.theme.bg.input};
        } // &__links
      } // @media only screen and (max-width : 576px)
    } // .header
  } // .container
`;

export const NavBar = styled.nav`
  display: flex;
  flex-grow: 1;
  justify-content: space-around;
  align-items: center;
  max-width: 700px;
`;
