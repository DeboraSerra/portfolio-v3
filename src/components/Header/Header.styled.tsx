import styled, { keyframes } from "styled-components";

const fadeIn = keyframes`
  0% {
    opacity: 0;
    transform: translateY(-10%);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
`;

const fadeLeft = keyframes`
  0% {
    opacity: 0;
    transform: translateX(-100%);
  }

  100% {
    opacity: 1;
    transform: translateX(0);
  }
`;

export const Header = styled.header`
  background-color: ${(props) => props.theme.bg.primary};
  padding: 20px 0;
  border-bottom: 1px solid ${(props) => props.theme.text.primary};
  height: 90px;

  color: ${(props) => props.theme.text.primary};

  nav {
    flex-basis: 70%;
  } // nav

  .hamburger-icon {
    font-size: ${(props) => props.theme.fontSizes[3]};
    cursor: pointer;
  } // .hamburger-icon

  .header {
    &.active {
      animation-name: ${fadeLeft};
      animation-duration: 0.5s;
    } // &.active

    &__logo {
      & img {
        border-radius: 50%;
      }// & img
    }// &__logo

    &__link {
      display: flex;
      gap: 2px;
      align-items: center;
      justify-content: flex-start;
      cursor: pointer;

      font-size: ${(props) => props.theme.fontSizes[3]};
      color: ${(props) => props.theme.text.tertiary};

      &:hover {
        color: inherit;
        text-decoration: underline;
      } // &:hover

      &:active {
        color: ${(props) => props.theme.bg.contrast[0]};
      } // &:active

      &.medium {
        font-size: ${(props) => props.theme.fontSizes[2]};
      } // &.medium

      &.small {
        font-size: ${(props) => props.theme.fontSizes[1]};
      } // &.small

      &.smallest {
        font-size: ${(props) => props.theme.fontSizes[0]};
      } // &.small
    } // &__link

    &__btn {
      padding: 12px 24px;
      border: none;
      border-radius: 4px;

      color: ${(props) => props.theme.bg.primary};
      background: transparent;

      color: ${(props) => props.theme.text.primary};
    } // &__btn

    &__menu {
      position: relative;

      &--abs {
        width: 180px;
        padding: 8px 12px;
        position: absolute;
        left: -12px;
        background-color: ${(props) => props.theme.bg.primary};
        box-shadow: 2px 4px 12px 4px ${(props) => props.theme.bg.shadow};
      } // &--abs
    } // &__menu

    @media only screen and (max-width: 768px) {
      position: absolute;
      min-width: 230px;
      top: 90px;
      background-color: ${(props) => props.theme.bg.primary};
      height: 100vh;
      left: 0;
      padding: 12px 24px;
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      gap: 24px;
      z-index: 3;

      &__menu {
        &--abs {
          position: static;
          transform: translateX(0);
          box-shadow: none;
          animation-name: ${fadeIn};
          animation-duration: 0.5s;
        } // &--abs
      } // &__menu
    } // @media only screen and (max-width : 768px)
  } // .header
`;
