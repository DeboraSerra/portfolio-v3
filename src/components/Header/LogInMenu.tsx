import { ProjectsContext } from "@/helpers/Context";
import { useContext } from "react";
import { BiChevronDown, BiChevronUp } from "react-icons/bi";
import ProfileMenu from "../ProfileMenu";

const LogInMenu = ({
  showProfileMenu,
  setShowProfileMenu,
}: {
  showProfileMenu: boolean;
  setShowProfileMenu: (val: boolean) => void;
}) => {
  const {
    user: { login },
    width,
  } = useContext(ProjectsContext);

  const isMobile = width <= 768;

  const renderArrow = () => {
    return showProfileMenu ? <BiChevronUp /> : <BiChevronDown />;
  };

  return (
    <div
      className='header__menu'
      onMouseEnter={() => !isMobile && login !== "" && setShowProfileMenu(true)}
      onMouseLeave={() =>
        !isMobile && login !== "" && setShowProfileMenu(false)
      }
      onClick={(e) => isMobile && login && e.preventDefault()}
    >
      <a
        className={`header__link ${isMobile ? "small" : "medium"}`}
        href={
          login
            ? "#"
            : `https://github.com/login/oauth/authorize?client_id=${process.env.NEXT_PUBLIC_GITHUB_CLIENT_ID}`
        }
        title={!login ? "Login with Github" : ""}
      >
        {login !== "" ? login : "Login"}
        {login !== "" ? renderArrow() : null}
      </a>
      {showProfileMenu ? <ProfileMenu /> : null}
    </div>
  );
};

export default LogInMenu;
