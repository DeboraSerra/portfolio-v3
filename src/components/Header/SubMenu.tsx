import { ProjectsContext } from "@/helpers/Context";
import Link from "next/link";
import { useContext } from "react";

const SubMenu = () => {
  const { routes, width } = useContext(ProjectsContext);
  const isMobile = width <= 768;
  return (
    <div className={`header__menu--abs ${!isMobile ? "flex column" : ""}`}>
      <Link
        className={`header__link ${isMobile ? "small" : "medium"}`}
        href={`/projects`}
      >
        All categories
      </Link>
      {routes.map(({ route, name, id }) => (
        <Link
          className={`header__link ${isMobile ? "small" : "medium"}`}
          key={id}
          href={`/projects/${route}`}
        >
          {name}
        </Link>
      ))}
    </div>
  );
};

export default SubMenu;
