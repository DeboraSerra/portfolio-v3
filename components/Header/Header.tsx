import Link from "next/link";
import { useRef, useState } from "react";
import { useDarkMode, useOnClickOutside, useWindowSize } from "usehooks-ts";
import { BiSun, BiMoon } from "react-icons/bi";
import { GiHamburgerMenu } from "react-icons/gi";
import { RxCross2 } from "react-icons/rx";

import * as S from "./Header.styled";
import Menu from "./components/Menu/Menu";

const Header = () => {
  const [openMenu, setOpenMenu] = useState(false);
  const { isDarkMode, toggle } = useDarkMode();
  const { width } = useWindowSize();
  const header = useRef<HTMLDivElement>(null);

  useOnClickOutside(header, () => setOpenMenu(false))

  return (
    <S.Header>
      <div className="container" ref={header}>
        <Link href="/" className="header__title">
          DÃ©bora Serra
        </Link>
        <S.NavBar>
          {width < 576 ? (
            <>
              {openMenu ? (
                <RxCross2
                  className="header__nav-btn"
                  onClick={() => setOpenMenu(!openMenu)}
                />
              ) : (
                <GiHamburgerMenu
                  className="header__nav-btn"
                  onClick={() => setOpenMenu(!openMenu)}
                />
              )}
              {openMenu && <Menu />}
            </>
          ) : (
            <Menu />
          )}
        </S.NavBar>
        <div onClick={toggle} className="header__theme-btn">
          {isDarkMode ? (
            <BiMoon className="header__theme-btn--icon" />
          ) : (
            <BiSun className="header__theme-btn--icon" />
          )}
        </div>
      </div>
    </S.Header>
  );
};

export default Header;
