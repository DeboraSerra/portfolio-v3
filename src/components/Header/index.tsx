import { ProjectsContext } from "@/helpers/Context";
import { useContext, useEffect, useRef, useState } from "react";
import { BiChevronDown, BiChevronUp } from "react-icons/bi";
import { BsMoonFill, BsSun } from "react-icons/bs";
import {
  useOnClickOutside,
  useTernaryDarkMode,
  useWindowSize,
} from "usehooks-ts";

import logo from "../../../assets/images/laptop.png";

import Image from "next/image";
import Link from "next/link";
import { useRouter } from "next/router";
import { GiHamburgerMenu } from "react-icons/gi";
import * as S from "./Header.styled";

const Header = () => {
  const { width } = useWindowSize();
  const header = useRef(null);
  const [showSubMenu, setShowSubMenu] = useState(false);
  const [showMenu, setShowMenu] = useState(false);
  useOnClickOutside(header, () => {
    setShowMenu(false);
  });
  const { routes } = useContext(ProjectsContext);
  const router = useRouter();

  const handleRouteChange = () => {
    setShowMenu(false);
    setShowSubMenu(false);
  };

  useEffect(() => {
    router.events.on("routeChangeStart", handleRouteChange);

    return () => {
      router.events.off("routeChangeStart", handleRouteChange);
    };
  }, [router.events]);

  const { isDarkMode, setTernaryDarkMode } = useTernaryDarkMode();

  return (
    <S.Header ref={header}>
      <div className='container flex'>
        {width <= 768 && (
          <GiHamburgerMenu
            onClick={() => setShowMenu(!showMenu)}
            className='hamburger-icon'
          />
        )}
        <Link href='/'>
          <Image src={logo} alt='DÃ©bora Serra' width={50} height={50} />
        </Link>
        {width > 768 ? (
          <nav className='flex center header'>
            <Link className='header__link medium' href='/'>
              Home
            </Link>
            <Link className='header__link medium' href='/resume'>
              Resume
            </Link>
            <div
              className='header__menu'
              onMouseEnter={() => setShowSubMenu(true)}
              onMouseLeave={() => setShowSubMenu(false)}
            >
              <Link className='header__link medium' href='/projects'>
                Projects
                {showSubMenu ? <BiChevronUp /> : <BiChevronDown />}
              </Link>
              {showSubMenu ? (
                <div className='header__menu--abs flex column'>
                  {routes.map(({ id, name, route }) => (
                    <Link
                      className='header__link medium'
                      key={id}
                      href={`/projects/${route}`}
                    >
                      {name}
                    </Link>
                  ))}
                </div>
              ) : null}
            </div>
            <Link className='header__link medium' href='/contact'>
              Contact
            </Link>
          </nav>
        ) : (
          showMenu && (
            <nav className='header'>
              <Link className='header__link small' href='/'>
                Home
              </Link>
              <Link className='header__link small' href='/resume'>
                Resume
              </Link>
              <div
                className='header__menu'
                onClick={() => setShowSubMenu(!showSubMenu)}
              >
                <Link
                  className='header__link small'
                  href='#!'
                  onClick={(e) => e.preventDefault()}
                >
                  Projects
                  {showSubMenu ? <BiChevronUp /> : <BiChevronDown />}
                </Link>
                {showSubMenu ? (
                  <div className='header__menu--abs'>
                    {routes.map(({ id, name, route }) => (
                      <Link
                        className='header__link small'
                        key={id}
                        href={`/projects/${route}`}
                      >
                        {name}
                      </Link>
                    ))}
                  </div>
                ) : null}
              </div>
              <Link className='header__link small' href='/contact'>
                Contact
              </Link>
            </nav>
          )
        )}
        <button
          className='header__btn'
          onClick={() => {
            isDarkMode
              ? setTernaryDarkMode("light")
              : setTernaryDarkMode("dark");
          }}
        >
          {isDarkMode ? <BsSun /> : <BsMoonFill />}
        </button>
      </div>
    </S.Header>
  );
};

export default Header;
