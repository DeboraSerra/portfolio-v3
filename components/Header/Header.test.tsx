import { render } from "@testing-library/react";
import TProvider from "../helpers/ThemeProvider";
import Header from "./Header";

describe("Test the Header component", () => {
  it("Should match the snapshot", () => {
    const { asFragment } = render(
      <TProvider>
        <Header />
      </TProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
