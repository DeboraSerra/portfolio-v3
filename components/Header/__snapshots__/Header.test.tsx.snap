// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Header component Should match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  background-color: #f1b2b2;
  padding: 12px 24px;
  margin-bottom: 12px;
  box-shadow: 0 4px 24px #45496a6f;
  height: 64px;
}

.c0 .container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .container .header__title {
  font-size: 32px;
  font-weight: 800;
  color: #45496a;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 .container .header__links {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 700px;
}

.c0 .container .header__link {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #2f3037;
  cursor: pointer;
}

.c0 .container .header__link:hover {
  color: #525560;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 .container .header__theme-btn,
.c0 .container .header__nav-btn {
  color: #2f3037;
  cursor: pointer;
  font-size: 22px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 700px;
}

@media only screen and (max-width:576px) {
  .c0 .container .header__links {
    position: absolute;
    top: 64px;
    left: 0;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: space-around;
    -webkit-justify-content: space-around;
    -ms-flex-pack: space-around;
    justify-content: space-around;
    width: 100%;
    background-color: #f1b2b2;
    padding: 12px 24px;
    box-shadow: 0 12px 24px #45496a6f;
  }
}

<header
    class="c0"
  >
    <div
      class="container"
    >
      <a
        class="header__title"
        href="/"
      >
        DÃ©bora Serra
      </a>
      <nav
        class="c1"
      >
        <div
          class="header__links"
        >
          <a
            class="header__link"
            href="/about"
          >
            About Me
          </a>
          <p
            class="header__link"
          >
            Projects 
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.293 9.293 12 13.586 7.707 9.293l-1.414 1.414L12 16.414l5.707-5.707z"
              />
            </svg>
          </p>
          <a
            class="header__link"
            href="/contact"
          >
            Contact
          </a>
        </div>
      </nav>
      <div
        class="header__theme-btn"
      >
        <svg
          class="header__theme-btn--icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.993 12c0 2.761 2.246 5.007 5.007 5.007s5.007-2.246 5.007-5.007S14.761 6.993 12 6.993 6.993 9.239 6.993 12zM12 8.993c1.658 0 3.007 1.349 3.007 3.007S13.658 15.007 12 15.007 8.993 13.658 8.993 12 10.342 8.993 12 8.993zM10.998 19h2v3h-2zm0-17h2v3h-2zm-9 9h3v2h-3zm17 0h3v2h-3zM4.219 18.363l2.12-2.122 1.415 1.414-2.12 2.122zM16.24 6.344l2.122-2.122 1.414 1.414-2.122 2.122zM6.342 7.759 4.22 5.637l1.415-1.414 2.12 2.122zm13.434 10.605-1.414 1.414-2.122-2.122 1.414-1.414z"
          />
        </svg>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
